velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,51 +28,117 @@ namespace facebook::velox::exec::prefixsort {
 /// Provides encode/decode methods for PrefixSort.
 class PrefixSortEncoder {
  public:
-  /// 1. Only int64_t is supported now.
-  /// 2. Encoding is compatible with sorting ascending with no nulls.
+  /// Encode native primitive types(such as byte, int16, int32, int64, double,
+  /// float) to dest. There will be a separate method to handle string type as
+  /// follow-up.
+  /// 1. The first byte of the encode result is nullByte, which is determined by
+  ///    the nullsFirst :
+  ///    -If nullsFirst : invalid(means null) -> 0, valid(means not null) -> 1
+  ///    -If !nullsFirst : invalid(means null) -> 1, valid(means not null) -> 0
+  /// 2. The remaining bytes are the encoding result of value:
+  ///    -If isNull, we ignore the input value (this means the caller
+  ///     can pass any value, the reason for not using std::optional here is to
+  ///     avoid the overhead of box-unboxing) and set the remaining sizeof(T)
+  ///     bytes to '0', they do not affect the comparison results at all.
+  ///    -If !isNull, we set the remaining bytes as the result of
+  ///     encodeNoNulls, which will consider whether it is ascending or not. The
+  ///     result in descending order is equal to invert-bits(the result in
+  ///     ascending order)
   template <typename T>
-  static FOLLY_ALWAYS_INLINE void encode(T value, char* row);
+  static FOLLY_ALWAYS_INLINE void
+  encode(bool ascending, bool nullsFirst, bool isNull, T value, char* dest);
+
+  /// 1. Only int64_t/int32_t is supported now.
+  /// 2. Assuming no nulls.
   template <typename T>
-  static FOLLY_ALWAYS_INLINE void decode(char* row);
+  static FOLLY_ALWAYS_INLINE void
+  encodeNoNulls(bool ascending, T value, char* dest);
+};
 
- private:
-  FOLLY_ALWAYS_INLINE static uint8_t flipSignBit(uint8_t byte) {
-    return byte ^ 128;
+template <typename T>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encode(
+    bool ascending,
+    bool nullsFirst,
+    bool isNull,
+    T value,
+    char* dest) {
+  const char valid = nullsFirst ? 1 : 0;
+  const char invalid = 1 - valid;
+  if (isNull) {
+    dest[0] = invalid;
+    simd::memset(dest + 1, 0, sizeof(T));
+  } else {
+    dest[0] = valid;
+    encodeNoNulls<T>(ascending, value, dest + 1);
   }
-};
+}
 
 /// Assuming that value is little-endian encoded, we encode it as follows to
-/// make sure encoding is compatible with sorting ascending:
+/// make sure encoding is compatible with memcmp-sorting ascending:
 /// 1 Reverse each byte, for example, 0xaabbccdd becomes 0xddccbbaa.
```

Comment:
This example is nice, but hard to read. Perhaps, add spaces between bytes:

`'0x aa bb cc dd' becomes '0x dd cc bb aa'`

This function takes 64-bit integer (8 bytes), but the example uses 4 bytes. Am I missing something? It would be nice to align the example with the code.